### ISSUE TITLE
Additional makefile rule for "runnable" examples

### ISSUE THREAD

**Issue by smiths** [JacquesCarette/Drasil#2149]
The makefile currently generates code that is missing dependencies, so the case studies will not compile and run right "out of the box."  To get the code to run requires copying over libraries and configuration files.  For instance, noPCM requires ODE solvers for each language and GlassBR requires the data files (TSD.txt and SDF.txt?).  The problem with copying the files over manually is that make will fail on the diff check because the local folders will not match stable.

From an e-mail exchange with @JacquesCarette, we need a new target in the Makefile, call it "runnable-nopcm" etc, that takes care of copying those files over in one go. That way you can also 'make clean' with less pain.

**Comment by balacij** [comment]
I started looking around on how this could be done, and I realized that this requires more work than I was expecting.

Presently,
1. We have a `SampleInput` data type that we use to copy over a *single* sample input file from the `datafiles/` folder when generating the source code files and SRS documents. This is currently insufficient for GlassBR -- as described in the original post, GlassBR requires TSD.txt and SDF.txt so those still need to be manually copied over.
2. If we want to run the examples, we have to manually copy over the dependencies for each.
3. We have the generated (/copied sampleInput.txt) `input.txt` files also inside of `stable/`.
4. For PDController, we expect dependencies to be installed in the system's python packages (scipy + numpy).

Ultimately, we need to be able to create/copy sample datafiles (do we?), and we need to be able to do some sort of dependency management.

#### Potential solutions for sample input files
1. Change or add a new variant of `SampleInput` so that we can copy over multiple files.
2. Remove `SampleInput` entirely and create symbolic links from `datafiles/<TSD&SDF&sampleInput&...>` into project directories.
3. Represent "SampleInputs" in the examples (I think this is the best solution). For each value that must be inputted into the program, we would give it a "default"/"sample" value and a file from which it orginates, and then we would allow the users of a program to `make genSampleData`/`<example> --genSampleData` to generate the sample data(files).

#### Potential solutions for dependency management
1. #402 (I think this is the best solution) -- If you've used Python's `virtualenv`, I think this will feel very similar to it in usage (not in setup or creation). It will allow our examples to be very non-invasive to user systems and allow for very easy software distribution.
2. Manually copying over files (hacky) -- storing dependencies in the `datafiles/` folder and copy them over (or use a symbolic link) before running.
3. Create a new variant of `SampleInput` to copy over dependencies too (hacky as well).

**Comment by smiths** [comment]
Great to see you analysis of this issue @balacij.  I definitely like the idea of having package and configuration management.  If someone forks Drasil, we would like them to be able to generate the case studies, and be able to run the generated code.

We should also think about how we would like Drasil to eventually be used.  My thinking is that the Drasil source for an example will be maintained by an "expert".  The expert will know the domain, and how to work with Drasil.  They will then use Drasil to generate the executable (or library) for their particular application.  The users of the generated executable (or library) wouldn't need to know Drasil.  They would just get the software delivered to them, without the Drasil infrastructure.  The expert would distribute software that looks like what people are currently generating manually.

Our goal, as I see it, is to generate software packages that are easy to install.  This is an ambitious goal, since there are so many ways that developers currently manage dependencies.  For our first pass, maybe we stick with generating a README file (or INSTALL file?) that lists the dependencies, but doesn't actually manage them?  Alternatively, we pick the "best" most portable package management solution and generate for that.  I'm not sure there is any clear winner on the best package management solution, especially when we consider different operating systems.

**Comment by JacquesCarette** [comment]
Regarding `SampleInput`: the best thing to do would be to generate these. Mucking around with files is fragile.  So potential solution 2 for sample input files is right out for me. Number 1 isn't great either. The third is definitely best, and would be better still if it were more generative.

For dependency management: nix is definitely the best solution overall, but solution 4 has its appeal as well. It needs exploration to see how much work it is, how invasive, etc. We need to understand what we'd be forcing on the end-users of Drasil-generated code.

**Comment by balacij** [comment]
Thank you both very much!

@smiths
> If someone forks Drasil, we would like them to be able to generate the case studies, and be able to run the generated code.

I agree! I hope these solutions we're approaching will do just that.

> This is an ambitious goal, since there are so many ways that developers currently manage dependencies. For our first pass, maybe we stick with generating a README file (or INSTALL file?) that lists the dependencies, but doesn't actually manage them? Alternatively, we pick the "best" most portable package management solution and generate for that. I'm not sure there is any clear winner on the best package management solution, especially when we consider different operating systems.

I see what you mean, it certainly is ambitious! We currently have the information about dependencies in the generated READMEs, but I think that we will need slightly more automation for the inputs for #2521 (and half of this ticket) to be done so that the generated artifacts can be more independent.


@JacquesCarette

Regarding `SampleInput`:
> The third is definitely best, and would be better still if it were more generative.

Sorry, I'm not quite sure what you mean by "would be better still if it were more generative". Do you have any ideas for how we can do this? I was thinking that this would need to be a part of the Choices that a user selects when wanting to generate the code using constants stored by the designers of the example.


I agree, I think that Nix would be the best first option since it has implementations for each OS (arguably excluding Windows which has to use WSL to use `nix` pkg manager). We should be able to add other package manager renderers when we've properly stored the information needed for generating the Nix files.

**Comment by JacquesCarette** [comment]
What I mean is that we should have more of a description of what's in `SampleInput`, and then generate the actual file. Though I don't really want to start putting too much information in the code either!  

Basically, I think we should minimize the amount of data 'outside Drasil' that is used to make artifacts. Of course, that depends on how we define 'inside' and 'outside'. Certainly any file that we just copy verbatim is 'outside' -- anything that isn't read and processed clearly doesn't contain information that **Drasil** manipulates (and is re-usable).

**Comment by smiths** [comment]
I agree with minimizing the data that is "outside Drasil."  We are a long way from getting there, but it would be nice if one day Drasil knew how to draw figures, because some changes in the models means changes to the figures.  Unfortunately right now we are obligated to change the figures manually to match the changes in the models.  If the data on the figures was inside Drasil, then the figures could be updated.  We like to promote how easy it is to change symbols in Drasil, but technically, it is only easy if the symbols don't appear in a figure.  :smile:

**Comment by balacij** [comment]
Sounds good.

It seems I will need to create a description of what's in SampleInput first, and then we can go from there for actionable steps.

(So, @Ant13731, this might need to be discussed a bit more first, I was too hasty! Sorry!)

**Comment by balacij** [comment]
I will similarly return to this later, but, if anyone wants to get to it first, please feel free to (unassigning for the time being)! I think the jist of the solution is to (i) create a chunk type (?) that describes general program types, (ii) describe a "sample" program configuration for (i), and then (iii) tie the sample configuration to a readable format that the generated programs can use.

**Comment by smiths** [comment]
Can we connect the sample input to the inputs listed in `Body.hs`.  For instance, for projectile, we have:

https://github.com/JacquesCarette/Drasil/blob/56bb4b73a3982aa905002c317ed3738364a8c77e/code/drasil-example/projectile/lib/Drasil/Projectile/Body.hs#L187-L188

Those are the same inputs that appear in the input file.  If our variables "know" their typical values, then I think we could generate the input file fairly easily.  Maybe typical value is already one of the fields that we store with a chunk?

**Comment by balacij** [comment]
I don't know why I didn't see your message earlier... sorry about that!

Yes, that could certainly work for sample inputs!
